src/java/org/apache/cassandra/db/memtable/pmem/PmemPartition.java

```diff
@@ -626,6 +616,158 @@ public Iterator<RangeTombstone> getMergedRangeTombstones(PartitionUpdate update)
+            {
+                artEntryKeyArray.add(nextEntry.getKey());
+            }
+        }
+        if (!artEntryKeyArray.isEmpty())
+        {
+            for (byte[] artEntryKey : artEntryKeyArray)
+            {
+                pmemRtmMap.getRangeTombstoneMarkerTree().remove(artEntryKey, (Long rowhandle) -> {
+                    TransactionalMemoryBlock rowMemoryBlock = heap.memoryBlockFromHandle(rowhandle);
+                    rowMemoryBlock.free();
+                });
+            }
+        }
+
+        return artEntryKeyArray.size();
+    }
+
+    private static int getGcBefore(TableMetadata metadata, int nowInSec)
+    {
+        ColumnFamilyStore cfs = Keyspace.open(metadata.keyspace).getColumnFamilyStore(metadata.id);
+        return cfs.gcBefore(nowInSec);
+    }
+
+    // Returns RTM deletion time
+    private static DeletionTime rtmDeletionTime(RangeTombstoneMarker marker)
+    {
+        return marker.isOpen(false) ? marker.openDeletionTime(false) : marker.closeDeletionTime(false);
+    }
+
+    private Unfiltered getUnfiltered(LongART.Entry nextEntry , boolean isRow)
+    {
+        Row.Builder builder = BTreeRow.sortedBuilder();
+        ByteComparable clusteringByteComparable = ByteComparable.fixedLength(nextEntry.getKey());
+        if(isRow)
+        {
+            Clustering<?> clustering = tableMetadata.comparator.clusteringFromByteComparable(ByteArrayAccessor.instance, clusteringByteComparable);
+            builder.newRow(clustering);
+        }
+        TransactionalMemoryBlock rangeMemoryBlock = heap.memoryBlockFromHandle(nextEntry.getValue());
+        DataInputPlus memoryBlockDataInputPlus = new MemoryBlockDataInputPlus(rangeMemoryBlock);
+        try
+        {
+            int savedVersion = (int) memoryBlockDataInputPlus.readUnsignedVInt();
+            SerializationHeader serializationHeader = pmemTableInfo.getSerializationHeader(savedVersion);
+            DeserializationHelper helper = new DeserializationHelper(metadata(), -1, DeserializationHelper.Flag.LOCAL);
+            return PmemRowSerializer.serializer.deserializeTombstone(memoryBlockDataInputPlus, serializationHeader, helper, builder);
+        }
+        catch (IOException e)
+        {
+            closeCartIterator();
+            throw new IOError(e);
+        }
+    }
+
+    /**
+     * True if unfiltered is expired otherwise false
+     * @param pmemTableInfo
+     * @param unfiltered
+     * @return
+     */
+    public static boolean isExpired(PmemTableInfo pmemTableInfo, Unfiltered unfiltered)
+    {
+        int localDeletionTime;
+        int gcBefore = getGcBefore(pmemTableInfo.getMetadata(),FBUtilities.nowInSeconds());
+        if (unfiltered.isRow())
+        {
+            Row row = (Row) unfiltered;
+            localDeletionTime = row.primaryKeyLivenessInfo().isExpiring() ? row.primaryKeyLivenessInfo().localExpirationTime() : row.deletion().time().localDeletionTime();
```

Comment:
What if there's a surviving cell in this row?

Note: the primary key liveness info only defines the lifetime of an empty row with no columns; the row must exist if live columns still exist.

Reply:
This check was being done to address expired TTLs. What you are saying makes sense, we have removed this check